polygeist/tools/cgeist/Lib/Attributes.cc

```diff
@@ -0,0 +1,329 @@
+        createNamedAttr(createStringAttr("nonnull", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Function does not recurse.
+  case llvm::Attribute::AttrKind::NoRecurse: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("norecurse", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Function does not return.
+  case llvm::Attribute::AttrKind::NoReturn: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("noreturn", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Parameter or return value may not contain uninitialized or poison bits.
+  case llvm::Attribute::AttrKind::NoUndef: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("noundef", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Function doesn't unwind stack.
+  case llvm::Attribute::AttrKind::NoUnwind: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("nounwind", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Function does not access memory.
+  case llvm::Attribute::AttrKind::ReadNone: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("readnone", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Function only reads from memory.
+  case llvm::Attribute::AttrKind::ReadOnly: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("readonly", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Hidden pointer to structure to return.
+  case llvm::Attribute::AttrKind::StructRet: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("sret", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Function always comes back to callsite.
+  case llvm::Attribute::AttrKind::WillReturn: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("willreturn", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  /// Function only writes to memory.
+  case llvm::Attribute::AttrKind::WriteOnly: {
+    NamedAttribute namedAttr =
+        createNamedAttr(createStringAttr("writeonly", dialect), unitAttr);
+    addAttribute(namedAttr);
+  } break;
+  default:
+    llvm_unreachable("Unexpected attribute kind");
+  }
+
+  return *this;
+}
+
+AttrBuilder &AttrBuilder::addAttribute(llvm::Attribute::AttrKind kind,
+                                       int64_t val) {
+  OpBuilder builder(&ctx);
+  constexpr StringLiteral dialect = LLVM::LLVMDialect::getDialectNamespace();
+
+  switch (kind) {
+  /// Alignment of parameter.
+  case llvm::Attribute::AttrKind::Alignment: {
+    assert(val > 0 && "Invalid alignment value");
+    assert(val <= 0x100 && "Alignment too large");
+    NamedAttribute namedAttr = createNamedAttr(
+        createStringAttr("align", dialect),
+        builder.getIntegerAttr(builder.getIntegerType(32), val));
```

Comment:
What is the reason to use a smaller integer type for this attribute? The smaller range of possible values?

Would it make sense to have the same integer type for all attributes, given that even Int32 still exceeds the possible range by far?

Reply:
I'd rather use `uint64_t` for all cases and for the signature, as this is what is used in all of [Attribute.h](https://github.com/intel/llvm/blob/44d7926c724c64c074bba55e50529a1cc25743fd/llvm/include/llvm/IR/Attributes.h#L708) signatures and that way we could check that the alignment is too large using `llvm::Value::MaximumAlignment`.